### DISCUSSION TITLE
Automatically format the import statements to adhere to the Google coding style guidelines

### DISCUSSION THREAD

**Discussion by yiliu30** [astral-sh/ruff#11343] [Q&A]
Can Ruff be used to format code in alignment with the guidelines outlined in https://google.github.io/styleguide/pyguide.html#22-imports?
> Use import statements for packages and modules only, not for individual types, classes, or functions.

For instance:
```python
# Original code
from module_name.util import helper_func1, helper_func2

helper_func1(a, b, ...)
helper_func2(a, ...)

# Expected code
from module_name import util

util.helper_func1(a, b, ...)
util.helper_func2(a, ...)
```

**Comment by tjkuson** [comment]
No, I don't think it can. If I understand the problem correctly, it would need multi-file analysis and better type inference to determine modules from other objects. Ruff can't do that yet (#3893, #7447).

You might be able to use [`banned-import-from` (`ICN003`)](https://docs.astral.sh/ruff/rules/banned-import-from/) to approximate the desired behaviour. For example, to lint against importing from `functools`, add

```toml
[tool.ruff.lint.flake8-import-conventions]
banned-from = ["functools"]
```

to your pyproject.toml file. This should report on `from functools import wraps` but not `import functools`.

**Reply by yiliu30** [reply]
@tjkuson  Thanks for your reply. I get that it's up to the users to specify which imports are banned.

A follow-up question, if I use [banned-import-from (ICN003)](https://docs.astral.sh/ruff/rules/banned-import-from/) to identify the banned imports, is possible to create a rule that can automatically correct these imports? For example, the logic of the rule may looks as below:


For example:
```
# pyproject.toml
[tool.ruff.lint.flake8-import-conventions]
banned-from = ["transformers"]
# auto-fix rule = xxxx
```

If found :
```python
from transformers import AutoModel, AutoTokenizer

model = AutoModel(...)
tokenizer = AutoTokenizer(...)
```
Replace these codes with:

```python
import transformers

model = transformers.AutoModel(...)
tokenizer = transformers.AutoTokenizer(...)

```

**Reply by tjkuson** [reply]
It's an open issue: #6372

I might work on it if I find the time later this week.
